### PR TITLE
Port setn, topn and adjuststack

### PR DESCRIPTION
**Pull request by noahgibbs**

### CODE REVIEW

File: yjit/src/codegen.rs [hunk 1]
```diff
@@ -993,6 +993,47 @@ fn gen_putobject_int2fix(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlo
...
+{
+    let nval:VALUE = jit_get_arg(jit, 0);
+    let VALUE(n) = nval;
+
+    let top_val:X86Opnd = ctx.stack_pop(0);
+    let dst_opnd:X86Opnd = ctx.stack_opnd(n.try_into().unwrap());
+    mov(cb, REG0, top_val);
+    mov(cb, dst_opnd, REG0);
+
+    let mapping = ctx.get_opnd_mapping(InsnOpnd::StackOpnd(0));
```
Comment (maximecb) [inline yjit/src/codegen.rs]: Let's put `use InsnOpnd::*` at the top of the file.

I also think the code would be more readable if n was converted into the right input type for stack_opnd earlier.

File: yjit/src/codegen.rs [hunk 2]
```diff
@@ -993,6 +993,47 @@ fn gen_putobject_int2fix(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlo
...
+    ctx.set_opnd_mapping(InsnOpnd::StackOpnd(n.try_into().unwrap()), mapping);
+
+    KeepCompiling
+}
+
+fn gen_topn(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, ocb: &mut CodeBlock) -> CodegenStatus
```
Comment (maximecb) [inline yjit/src/codegen.rs]: There was a comment above this function describing what this instruction does. Let's try to preserve the comments.

File: yjit/src/codegen.rs [hunk 3]
```diff
@@ -993,6 +993,47 @@ fn gen_putobject_int2fix(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlo
...
+    let mapping = ctx.get_opnd_mapping(InsnOpnd::StackOpnd(n.try_into().unwrap()));
+
+    let loc0 = ctx.stack_push_mapping(mapping);
+    mov(cb, REG0, top_n_val);
+    mov(cb, loc0, REG0);
+
+    KeepCompiling
+}
+
+fn gen_adjuststack(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, ocb: &mut CodeBlock) -> CodegenStatus
```
Comment (maximecb) [inline yjit/src/codegen.rs]: Comment above this function should be preserved here too.

**Review by maximecb — CHANGES_REQUESTED** [review]
Left some comments with small requested changes.

File: yjit/src/codegen.rs (hunk 1 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs]: We already have a use InsnOpnd::* in the "use" section up-top. Is that what you're thinking of, or am I misunderstanding?

Agree about the conversion, I'll fix it.

File: yjit/src/codegen.rs (hunk 2 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs]: Okay. I've restored the comments and fixed the type conversion.

File: yjit/src/codegen.rs (hunk 1 above)
Comment (maximecb) [reply yjit/src/codegen.rs]: What I mean is that if we do `use InsnOpnd::*` then we can just go `StackOpnd(0)` without the prefix. Makes it slightly shorter and more readable.

And I just noticed that you need both n as u16 and u32. That's something I should fix.

Comment (maximecb) [reply yjit/src/codegen.rs]: With respect to the type conversion, I think you had the right approach. I apologize.

File: yjit/src/codegen.rs (hunk 2 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs]: Okay. I've changed the type conversion back, and removed that prefix in a few places we were using it.

**Comment by noahgibbs** [comment]
Looks like this now has conflicts. I'll rebase and re-push.

**Comment by noahgibbs** [comment]
Rebased to merge and pass tests with new changes.

**Comment by maximecb** [comment]
Looking good 👌 :)

**Pull request merged**